### PR TITLE
WFLY-18066 ByteBufferMarshalledValue generates duplicate buffers during a single marshalling operation

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.redhat.com/browse/WFLY-18066

### CODE REVIEW

File: clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java [hunk 1]
```diff
@@ -64,11 +65,18 @@ public synchronized boolean isEmpty() {
...
+            buffer = this.marshaller.write(this.object);
+            // N.B. Refrain from logging wrapped object
+            // If wrapped object contains an EJB proxy, toString() will trigger an EJB invocation!
+            LOGGER.debugf("Marshalled size of %s object = %d bytes", this.object.getClass().getCanonicalName(), buffer.limit() - buffer.arrayOffset());
+        }
+        return buffer;
+    }
+
+    public synchronized OptionalInt size() {
+        return (this.buffer != null) ? OptionalInt.of(this.buffer.limit() - this.buffer.arrayOffset()) : this.marshaller.size(this.object);
```
Comment (bstansberry) [inline clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java:79]: A comment explaining this algorithm would be good. It's not clear why passing a buffer into the constructor (so this.buffer != null) means the state of that buffer means this.buffer.arrayOffset() is safe to call (i.e. it's a heap buffer) or that this.buffer.limit() - this.buffer.arrayOffset() is the correct value when size() is called.

**Review by bstansberry — CHANGES_REQUESTED** [review]

File: clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java (hunk 1 above)
Comment (pferraro) [reply clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java:79]: Valid point.  In practice, these are always array backed buffers - but we should always have hasArray() checks (and alternate handling) for completeness.

**Comment by pferraro** [comment]
Added hold label while I finish the unit test.

File: clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java (hunk 1 above)
Comment (pferraro) [reply clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java]: Fixed.

Comment (pferraro) [reply clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java]: Oh - I see what you mean now.  Yes - I will add comments clarifying the lifecycle of this property, which is invalidated once the object is restored from the buffer.

Comment (pferraro) [reply clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java]: Done.

**Review by bstansberry — APPROVED** [review]
Thanks, @pferraro -- the comments are very helpful.

**Comment by bstansberry** [comment]
Thanks @pferraro

**Pull request merged**